risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+        7868944258580147481,
+        14744321562856667967,
+        2513567076326282710,
+        5089696809409609209,
+        17260140776825220475,
+        11898519751787946856,
+        15307271466853436433,
+        5456584715443070302,
+        1219213613525454263,
+        13843946492009319323,
+        16884827967813875098,
+        10516896061424301529,
+        4514835231089717636,
+        16488041148801377373,
+        16303955383020744715,
+        10790884855407511297,
+        8554224884056360729,
+    ];
+}
+
+impl Elem {
+    /// Create a new [Goldilocks] from a raw integer.
+    pub const fn new(x: u64) -> Self {
+        Self(x % P)
+    }
+}
+
+impl ops::Add for Elem {
+    type Output = Self;
+    /// Addition for Goldilocks field elements [Elem]
+    fn add(self, rhs: Self) -> Self {
+        Elem(add(self.0, rhs.0))
+    }
+}
+
+impl ops::AddAssign for Elem {
+    /// Simple addition case for Goldilocks field elements [Elem]
+    fn add_assign(&mut self, rhs: Self) {
+        self.0 = add(self.0, rhs.0)
+    }
+}
+
+impl ops::Sub for Elem {
+    type Output = Self;
+    /// Subtraction for Goldilocks field elements [Elem]
+    fn sub(self, rhs: Self) -> Self {
+        Elem(sub(self.0, rhs.0))
+    }
+}
+
+impl ops::SubAssign for Elem {
+    /// Simple subtraction case for Goldilocks [Elem]
+    fn sub_assign(&mut self, rhs: Self) {
+        self.0 = sub(self.0, rhs.0)
+    }
+}
+
+impl ops::Mul for Elem {
+    type Output = Self;
+    /// Multiplication for Goldilocks [Elem]
+    fn mul(self, rhs: Self) -> Self {
+        Elem(mul(self.0, rhs.0))
+    }
+}
+
+impl ops::MulAssign for Elem {
+    /// Simple multiplication case for Goldilocks [Elem]
+    fn mul_assign(&mut self, rhs: Self) {
+        self.0 = mul(self.0, rhs.0)
+    }
+}
+
+impl ops::Neg for Elem {
+    type Output = Self;
+    /// Negation for Goldilocks [Elem]
+    fn neg(self) -> Self {
+        Elem(0) - self
+    }
+}
+
+impl From<&Elem> for u64 {
+    fn from(x: &Elem) -> Self {
+        x.0
+    }
+}
+
+impl From<Elem> for u64 {
+    fn from(x: Elem) -> Self {
+        x.0.into()
+    }
+}
+
+impl From<u64> for Elem {
+    fn from(x: u64) -> Self {
+        Elem(x % P)
+    }
+}
+
+/// Wrapping addition of [Elem] using Goldilocks prime modulus
+fn add(lhs: u64, rhs: u64) -> u64 {
+    let x = lhs.wrapping_add(rhs);
+    // If we've done an u64
+    if x < rhs || x >= P {
+        x.wrapping_sub(P)
+    } else {
+        x
+    }
+}
+
+/// Wrapping subtraction of [Elem] using Goldilocks prime
+fn sub(lhs: u64, rhs: u64) -> u64 {
+    let x = lhs.wrapping_sub(rhs);
+    return if x > lhs { x.wrapping_add(P) } else { x };
+}
+
+/// Wrapping multiplication of [Elem] using Goldilocks prime
+fn mul(lhs: u64, rhs: u64) -> u64 {
+    // To prevent u64 overflow, we first perform with u128
+    let prod: u128 = (lhs as u128).wrapping_mul(rhs as u128);
+
+    // Initialize result to low 64 bits (by converting)
+    let mut ret: u64 = prod as u64;
```

Comment:
In rust, you can 'shadow' variables without ambiguity. So instead of using a `mut`, you could just create a new lvalue for each.

For example, you could say:
```
let mut x = 1;
x = x + 1;
```

Or:
```
let x = 1;
let x = x + 1;
```